File: typescript-examples/computer-use/api/anthropic-computer-use.ts
import z from "zod";
import { BrowserContext, Page } from "playwright";
import { getAiGatewayConfig } from "@intuned/runtime";
import { samplingLoop } from "../lib/anthropic/sampling-loop";

interface Params {
  query: string;  // The task you want the AI to perform
}

export default async function handler(
  params: Params,
  page: Page,
  _: BrowserContext,
) {
  const { query } = params;

  if (!query) {
    throw new Error('Query is required');
  }

  // Get AI gateway config
  const { apiKey, baseUrl } = await getAiGatewayConfig();

  // Hardcoded configuration
  const model = 'claude-sonnet-4-5';
  const thinkingBudget = 1024;
  const maxTokens = 4096;

  // Set viewport size to match the computer tool's display dimensions
  await page.setViewportSize({
    width: 1280,
    height: 720,
  });

  try {
    // Run the sampling loop
    const finalMessages = await samplingLoop({
      model,
      messages: [{
        role: 'user',
        content: query,
      }],
      apiKey,
      baseURL: baseUrl,
      thinkingBudget,
      maxTokens,
      playwrightPage: page,
    });

    // Extract the final result from the messages
    if (finalMessages.length === 0) {
      throw new Error('No messages were generated during the sampling loop');
    }

    const lastMessage = finalMessages[finalMessages.length - 1];
    if (!lastMessage) {
      throw new Error('Failed to get the last message from the sampling loop');
    }

    const result = typeof lastMessage.content === 'string'
      ? lastMessage.content
      : lastMessage.content.map(block =>
        block.type === 'text' ? block.text : ''
      ).join('');

    return {
      result,
      messageCount: finalMessages.length,
    };
  } catch (error) {
    console.error('Error in sampling loop:', error);
    throw error;
  }
}

File: python-examples/computer-use/api/anthropic-computer-use.py
from typing import TypedDict

from intuned_runtime import get_ai_gateway_config
from lib.anthropic.utils.loop import sampling_loop
from playwright.async_api import Page


class Params(TypedDict):
    query: str  # The task you want the AI to perform


async def automation(page: Page, params: Params | None = None, **_kwargs):
    if not params or not params.get("query"):
        raise ValueError("Query is required, please provide a query in the params")

    # Get AI gateway config
    base_url, api_key = get_ai_gateway_config()

    # Hardcoded model
    model = "claude-haiku-4-5"

    # Set viewport size to match the computer tool's display dimensions
    await page.set_viewport_size({"width": 1280, "height": 720})

    final_messages = await sampling_loop(
        model=model,
        messages=[
            {
                "role": "user",
                "content": params["query"],
            }
        ],
        api_key=api_key,
        base_url=base_url,
        thinking_budget=1024,
        playwright_page=page,
    )

    # Extract the final result
    if not final_messages:
        raise ValueError("No messages were generated during the sampling loop")

    last_message = final_messages[-1]
    if not last_message:
        raise ValueError("Failed to get the last message from the sampling loop")

    result = ""
    if isinstance(last_message.get("content"), str):
        result = last_message["content"]  # type: ignore[assignment]
    else:
        result = "".join(
            block["text"]
            for block in last_message["content"]  # type: ignore[index]
            if isinstance(block, dict) and block.get("type") == "text"
        )

    return {"result": result}
